sweep/sweeper.go

```diff
@@ -30,13 +63,671 @@ type UtxoSweeperConfig struct {
+	if !atomic.CompareAndSwapUint32(&s.stopped, 0, 1) {
+		return nil
 	}
+
+	log.Tracef("Sweeper shutting down")
+
+	close(s.quit)
+	s.wg.Wait()
+
+	log.Tracef("Sweeper shut down")
+
+	return nil
+}
+
+// SweepInput sweeps inputs back into the wallet. The inputs will be batched and
+// swept after the batch time window ends.
+//
+// NOTE: Extreme care needs to be taken that input isn't changed externally.
+// Because it is an interface and we don't know what is exactly behind it, we
+// cannot make a local copy in sweeper.
+func (s *UtxoSweeper) SweepInput(input Input) (chan Result, error) {
+	if input == nil || input.OutPoint() == nil || input.SignDesc() == nil {
+		return nil, errors.New("nil input received")
+	}
+
+	log.Debugf("Sweep request received: %v", input.OutPoint())
+
+	sweeperInput := &sweepInputMessage{
+		input:      input,
+		resultChan: make(chan Result, 1),
+	}
+
+	// Deliver input to main event loop.
+	select {
+	case s.newInputs <- sweeperInput:
+	case <-s.quit:
+		return nil, fmt.Errorf("sweeper shutting down")
+	}
+
+	return sweeperInput.resultChan, nil
+}
+
+// collector is the sweeper main loop. It processes new inputs, spend
+// notifications and counts down to publication of the sweep tx.
+func (s *UtxoSweeper) collector() {
+	defer s.wg.Done()
+
+	for {
+		select {
+		case input := <-s.newInputs:
+			outpoint := *input.input.OutPoint()
+			pendInput, pending := s.pendingInputs[outpoint]
+			if pending {
+				log.Debugf("Already pending input %v received",
+					outpoint)
+
+				// Add additional result channel to signal spend
+				// of this input.
+				pendInput.listeners = append(
+					pendInput.listeners, input.resultChan)
+
+				continue
+			}
+
+			// Create a new pendingInput and initialize the
+			// listeners slice with the passed in result channel. If
+			// this input is offered for sweep again, the result
+			// channel will be appended to this slice.
+			pendInput = &pendingInput{
+				listeners: []chan Result{input.resultChan},
+				input:     input.input,
+				state:     stateNew,
+			}
+			s.pendingInputs[outpoint] = pendInput
+
+			size, err := getInputWitnessSizeUpperBound(input.input)
+			if err != nil {
+				err = fmt.Errorf("cannot determine size: %v",
+					err)
+
+				s.signalAndRemove(&outpoint, Result{Err: err})
+
+				continue
+			}
+			pendInput.witnessSizeUpperBound = size
+
+			// Start watching for spend of this input, either by us
+			// or the remote party. The possibility of a remote
+			// spend is the reason why just registering for conf. of
+			// our sweep tx isn't enough.
+			cancel, err := s.registerSpendNtfn(input.input)
+			if err != nil {
+				err = fmt.Errorf("cannot watch for spend: %v",
+					err)
+
+				s.signalAndRemove(&outpoint, Result{Err: err})
+
+				continue
+			}
+			pendInput.ntfnRegCancel = cancel
+
+			// Lookup if there is already a confirmed or unconfirmed
+			// tx of ours that spends this input.
+			spendingTx, err := s.cfg.HasSpendingTx(outpoint)
```

Comment:
Should be replaced with a query to the local input DB?

Reply:
Yes, so we don't lookup this anymore, but blindly keep trying to get txes in.